Lib/test/test_builtin.py

```diff
@@ -2485,6 +2486,17 @@ def test_type_qualname(self):
             A.__qualname__ = b'B'
         self.assertEqual(A.__qualname__, 'D.E')
 
+    def test_type_typeparams(self):
+        class A[T]:
+            pass
+        T, = A.__type_params__
+        self.assertIsInstance(T, typing.TypeVar)
+        A.__type_params__ = "whatever"
```

Comment:
For functions we error if `__type_params__` is set to a non-tuple. Why not the same for types?

Reply:
We can't prevent people anyway from setting it to something random if they go directly into the `__dict__`, so I figured a type check here didn't make much sense.

I noticed that `function.__doc__` will happily let you set a function's docstring to an int, so maybe we should just also remove the typecheck there.

Reply:
> We can't prevent people anyway from setting it to something random if they go directly into the __dict__, so I figured a type check here didn't make much sense.

The `__dict__` of a type is a read-only `MappingProxyType`, so I don't think people can set it directly.

I don't have super strong opinions about whether we let people set it to nonsense values, but it's easier to relax restrictions later than to tighten them.

Reply:
```
Python 3.12.0a7+ (heads/main-dirty:dbe171e609, May 19 2023, 08:21:26) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class C[T]: pass
...
>>> C.__dict__
mappingproxy({'__module__': '__main__', '__type_params__': (T,), '__orig_bases__': (typing.Generic[T],), '__dict__': <attribute '__dict__' of 'C' objects>, '__weakref__': <attribute '__weakref__' of 'C' objects>, '__doc__': None, '__parameters__': (T,)})
>>> C.__dict__["__type_params__"] = "foo"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'mappingproxy' object does not support item assignment
```

Reply:
Jelle means this:

```pycon
>>> class Foo[T]:
...     __type_params__ = 'no introspection for YOU!'
>>> Foo.__type_params__
'no introspection for YOU!'
```

Reply:
Oh, I see, makes sense!